File: ProjectTrackervb/ProjectTracker.Library/Admin/Role.vb
Namespace Admin

  <Serializable()> _
  Public Class Role
    Inherits BusinessBase(Of Role)

#Region " Business Methods "

    Private Shared IdProperty As PropertyInfo(Of Integer) = RegisterProperty(Of Integer)(GetType(Role), New PropertyInfo(Of Integer)("Id"))
    Public Property Id() As Integer
      Get
        Dim result = GetProperty(Of Integer)(IdProperty)
        If result = 0 Then
          result = GetMax() + 1
          LoadProperty(Of Integer)(IdProperty, result)
        End If
        Return result
      End Get
      Set(ByVal value As Integer)
        SetProperty(Of Integer)(IdProperty, value)
      End Set
    End Property

    Private Function GetMax() As Integer

      ' generate a default id value
      Dim parent As Roles = CType(Me.Parent, Roles)
      Dim max As Integer = 0
      For Each item As Role In parent
        If Not ReferenceEquals(item, Me) AndAlso item.Id > max Then
          max = item.Id
        End If
      Next
      Return max

    End Function

    Private Shared NameProperty As PropertyInfo(Of String) = RegisterProperty(Of String)(GetType(Role), New PropertyInfo(Of String)("Name"))
    Public Property Name() As String
      Get
        Return GetProperty(NameProperty)
      End Get
      Set(ByVal value As String)
        SetProperty(Of String)(NameProperty, value)
      End Set
    End Property

    Private _timestamp(7) As Byte

#End Region

#Region " Validation Rules "

    Protected Overrides Sub AddBusinessRules()

      ValidationRules.AddRule(Of Role)( _
        AddressOf NoDuplicates, IdProperty)
      ValidationRules.AddRule( _
        AddressOf Csla.Validation.CommonRules.StringRequired, NameProperty)

    End Sub

    Private Shared Function NoDuplicates(Of T As Role)( _
      ByVal target As T, _
      ByVal e As Csla.Validation.RuleArgs) As Boolean

      Dim parent As Roles = CType(target.Parent, Roles)
      If parent IsNot Nothing Then
        For Each item As Role In parent
          If item.Id = target.GetProperty(Of Integer)(IdProperty) AndAlso Not ReferenceEquals(item, target) Then
            e.Description = "Role Id must be unique"
            Return False
          End If
        Next
      End If
      Return True

    End Function

#End Region

#Region " Authorization Rules "

    Protected Overrides Sub AddAuthorizationRules()

      AuthorizationRules.AllowWrite(IdProperty, "Administrator")
      AuthorizationRules.AllowWrite(NameProperty, "Administrator")

    End Sub

#End Region

#Region " Factory Methods "

    Friend Shared Function NewRole() As Role

      Return DataPortal.CreateChild(Of Role)()

    End Function

    Friend Shared Function GetRole(ByVal data As ProjectTracker.DalLinq.getRolesResult) As Role

      Return DataPortal.FetchChild(Of Role)(data)

    End Function

    Private Sub New()

      ' require use of factory methods

    End Sub

#End Region

#Region " Data Access "

    Private Sub Child_Fetch(ByVal data As ProjectTracker.DalLinq.getRolesResult)

      LoadProperty(Of Integer)(IdProperty, data.Id)
      LoadProperty(Of String)(NameProperty, data.Name)
      _timestamp = data.LastChanged.ToArray

    End Sub

    Private Sub Child_Insert()

      Using mgr = ContextManager(Of ProjectTracker.DalLinq.PTrackerDataContext).GetManager("PTracker", True)
        Dim lastChanged As System.Data.Linq.Binary = _timestamp
        mgr.DataContext.addRole(GetProperty(Of Integer)(IdProperty), _
                                GetProperty(Of String)(NameProperty), _
                                lastChanged)
        _timestamp = lastChanged.ToArray
      End Using

    End Sub

    Private Sub Child_Update()

      Using mgr = ContextManager(Of ProjectTracker.DalLinq.PTrackerDataContext).GetManager("PTracker", True)
        Dim lastChanged As System.Data.Linq.Binary = Nothing
        mgr.DataContext.UpdateRole(GetProperty(Of Integer)(IdProperty), _
                                   GetProperty(Of String)(NameProperty), _
                                   _timestamp, _
                                   lastChanged)
        _timestamp = lastChanged.ToArray
      End Using

    End Sub

    Private Sub Child_DeleteSelf()

      Using mgr = ContextManager(Of ProjectTracker.DalLinq.PTrackerDataContext).GetManager("PTracker", True)
        mgr.DataContext.DeleteRole(Id)
      End Using

    End Sub

#End Region

  End Class

End Namespace

File: ProjectTrackercs/ProjectTracker.Library/Admin/Role.cs
using Csla;
using Csla.Data;
using System;

namespace ProjectTracker.Library
{
  namespace Admin
  {

    [Serializable()]
    public class Role : BusinessBase<Role>
    {
      #region  Business Methods

      private static PropertyInfo<int> IdProperty = 
        RegisterProperty(new PropertyInfo<int>("Id"));
      private bool _idSet;
      public int Id
      {
        get
        {
          if (!_idSet)
          {
            _idSet = true;
            SetProperty(IdProperty, GetMax() + 1);
          }
          return GetProperty(IdProperty);
        }
        set
        {
          _idSet = true;
          SetProperty(IdProperty, value);
        }
      }

      private int GetMax()
      {
        // generate a default id value
        Roles parent = (Roles)this.Parent;
        int max = 0;
        foreach (Role item in parent)
        {
          if (item.Id > max)
          {
            max = item.Id;
          }
        }
        return max;

      }

      private static PropertyInfo<string> NameProperty = RegisterProperty(typeof(Role), new PropertyInfo<string>("Name"));
      public string Name
      {
        get
        {
          return GetProperty(NameProperty);
        }
        set
        {
          SetProperty(NameProperty, value);
        }
      }

      private byte[] _timestamp = new byte[8];

      #endregion

      #region  Validation Rules

      protected override void AddBusinessRules()
      {
        ValidationRules.AddRule<Role>(NoDuplicates, IdProperty);
        ValidationRules.AddRule(
          Csla.Validation.CommonRules.StringRequired, NameProperty);
      }

      private static bool NoDuplicates<T>(
        T target, Csla.Validation.RuleArgs e) where T : Role
      {
        Roles parent = (Roles)target.Parent;
        if (parent != null)
        {
          foreach (Role item in parent)
          {
            if (item.Id == target.ReadProperty(IdProperty) && 
              !(ReferenceEquals(item, target)))
            {
              e.Description = "Role Id must be unique";
              return false;
            }
          }
        }
        return true;
      }

      #endregion

      #region  Authorization Rules

      protected override void AddAuthorizationRules()
      {
        AuthorizationRules.AllowWrite(IdProperty, "Administrator");
        AuthorizationRules.AllowWrite(NameProperty, "Administrator");
      }

      #endregion

      #region  Factory Methods

      internal static Role NewRole()
      {
        return DataPortal.CreateChild<Role>();
      }

      internal static Role GetRole(ProjectTracker.DalLinq.getRolesResult data)
      {
        return DataPortal.FetchChild<Role>(data);
      }

      private Role()
      { /* require use of factory methods */ }

      #endregion

      #region  Data Access

      private void Child_Fetch(ProjectTracker.DalLinq.getRolesResult data)
      {
        LoadProperty(IdProperty, data.Id);
        _idSet = true;
        LoadProperty(NameProperty, data.Name);
        _timestamp = data.LastChanged.ToArray();
      }

      private void Child_Insert()
      {
        using (var mgr = ContextManager<ProjectTracker.DalLinq.PTrackerDataContext>.GetManager(ProjectTracker.DalLinq.Database.PTracker))
        {
          System.Data.Linq.Binary lastChanged = _timestamp;
          mgr.DataContext.addRole(ReadProperty(IdProperty), ReadProperty(NameProperty), ref lastChanged);
          _timestamp = lastChanged.ToArray();
        }
      }

      private void Child_Update()
      {
        using (var mgr = ContextManager<ProjectTracker.DalLinq.PTrackerDataContext>.GetManager(ProjectTracker.DalLinq.Database.PTracker))
        {
          System.Data.Linq.Binary lastChanged = null;
          mgr.DataContext.updateRole(ReadProperty(IdProperty), ReadProperty(NameProperty), _timestamp, ref lastChanged);
          _timestamp = lastChanged.ToArray();
        }
      }

      private void Child_DeleteSelf()
      {
        using (var mgr = ContextManager<ProjectTracker.DalLinq.PTrackerDataContext>.GetManager(ProjectTracker.DalLinq.Database.PTracker))
        {
          mgr.DataContext.deleteRole(Id);
        }
      }

      #endregion

    }
  }
}